Issues/reader card updates

aerych opened the pull request:
Closes #5762 
Closes #5763
Closes #5769
Refs #6070

This PR updates the reader cards to match the new default cards being rolled out to the calypso reader.  The refreshed card styles are documented on an internal blog. Ping me for a link.  The changes supersede many of the requests in #5769 

To test:
Check the post list and confirm the card layout is correct and that there are no breaking constraints warnings in the console.
Confirm that follow, share, like, and comment buttons have text labels on wide screens (iPad, landscape iPhone) and do not have labels on small screens (portrait iPhone, multitasking iPad). 
Confirm that:
- the follow button toggles following status and reflects the correct status.
- the share button opens the share sheet
- the visit button opens the post in a browser
- the tag button continues to preview a tag. 
- the tab button highlights correctly when tapped

Note that for the tag button we're using the same trick as the blog name label and positioning a button with a larger tap area above the tag label to make it easier for the user to interact with it.

Additionally, the follow button in the tag and site headers is updated to match the new style.  Confirm these buttons continue to reflect the correct state. 

Needs review: @kurzee could I trouble you with this one?

cc @nbradbury @jancavan

### Examples
![simulator screen shot nov 9 2016 7 19 14 pm](https://cloud.githubusercontent.com/assets/1435271/20181211/dfce11fc-a722-11e6-9833-2a1789792150.png)

![simulator screen shot nov 9 2016 7 25 36 pm](https://cloud.githubusercontent.com/assets/1435271/20181230/ecf63c4c-a722-11e6-9c77-e5df8803b6aa.png)

jancavan commented:
Nice work @aerych! Below are some feedback:

- Action button labels should be the same color as the icons themselves. Both should be `greyLighten10` / `#a8bece`:

![screenshot 2016-11-10 10 49 40](https://cloud.githubusercontent.com/assets/4924246/20189736/67c7c714-a733-11e6-8abb-5ecfdf0618b3.png)

- Align Visit button with the rest of the card content:
![screenshot 2016-11-10 11 48 25](https://cloud.githubusercontent.com/assets/4924246/20191715/9acc2558-a73b-11e6-85bf-f78d34f87a5b.png)

- Timestamp color should also be `greyLighten10` / `#a8bece`.

- In Calypso, Timestamps are in this format:
`2d ago`, `6d ago` — so just `d` instead of `Days`

- Timestamp should link to the external site post instead of their Reader site stream.

- Follow/Following icon and labels can use a bit of space in between:
![screenshot-2016-11-10-10 54 20](https://cloud.githubusercontent.com/assets/4924246/20190000/7499b474-a734-11e6-8c8a-5c1689bba75c.jpg)

- Follow/Following buttons should not exist in the Following stream since you're following these sites already. They should only be in other streams/tags/lists.

- Doesn't seem like you can actually share/reblog the post?

- For the `PostOptions` ellipsis, we don't have Visit — only Unfollow and Edit.

- Similar to @fraying's feedback, in default cards, images are `100px` tall. They're only taller in photo-cards (`225px`) which I know we have yet.

- Remove placeholder grav/blav if a site doesn't have any of the two and just scoot the meta info over to the left:
![screenshot-2016-11-10-11 20 11](https://cloud.githubusercontent.com/assets/4924246/20190836/ecd7fbfa-a737-11e6-9a0b-01b15a6e163e.jpg)

fraying commented:
Thanks, Jan! Agreed on all.

RE: the tag. Let's drop the tag and link all of the header stuff to the site view. This, I believe, is what both apps do now. Let’s just replicate that and move on the the much more important other parts of the refresh. We can circle back to adding the tag in a different way when the rest is in place. 

So, don't worry about linking the date to the post. Continue what you're doing now, linking it to the site view along with the avatar, name, and site. Cool?

kurzee commented:
When I tap on a tag in the header, the highlighted state sticks, and does not unhighlight unless recycled.

kurzee commented:
This was probably already discussed elsewhere, but the "More" button or "•••" button, is fairly redundant. The primary awkwardness though, is that the Share icon seems out of place on native as most native iOS apps instead default to using the "•••" for opening the native share sheet.

I'd recommend we remove the share icon/button and instead allow the "•••" button to open the share sheet. We already have "Visit" as its own button anyways.

kurzee commented:
The Follow icon in the top-right of the header seems crazy small and not obvious. It also seems redundant to see the follow icon for sites I already follow.

kurzee commented:
In my honest opinion, the visit, share, comment, like, •••, follow, and ~~tag~~ buttons on this refresh all feel like they were designed for a mouse pointer and not a finger touch on mobile. I'm afraid to scroll, in that I might accidentally knick one of the buttons with my finger.

The many buttons probably work great on the web, but on mobile this feels like we're trying to cram as many actionable items into each card as possible, taking away from the card content itself and not taking advantage of mobile norms to wrap interaction into the expect "•••" buttons.

kurzee commented:
@aerych it looks like we now need to set the height anchor for the `actionStackView` in `ReaderPostCardCell.xib` to have a required priority @ `1000`. Previously it was `999` for reasons I can't recall. I have a bug I can replicate on private sites that this would fix, DM for steps if needed.

kurzee commented on WordPress/Classes/ViewRelated/Reader/ReaderPostCardCell.swift:
```diff
@@ -388,9 +440,42 @@ import WordPressShared
+        let likeCount = provider.likeCount().integerValue
+        let commentCount = provider.commentCount().integerValue
+
+        if superview?.frame.width < 480 {
```
Maybe `contentView.traitCollection.horizontalSizeClass == .Compact`?

aerych replied:
I tried this but it seems unreliable.  For example, landscape orientation on an iPhone 6s still shows just an icon instead of an icon + title.

kurzee commented on WordPress/Classes/ViewRelated/Reader/ReaderPostCardCell.swift:
```diff
@@ -460,6 +566,14 @@ import WordPressShared
+    @IBAction func tagButtonTouchesDidEnd(sender: UIButton) {
+        tagLabelButton.highlighted = false
+    }
+
```
Probably just need to swap "Touch Up Inside" to trigger `tagButtonTouchesDidEnd` instead of `tagButtonTouchesDidHighlight` to resolve https://github.com/wordpress-mobile/WordPress-iOS/pull/6188#issuecomment-260431117

jancavan commented:
> I'd recommend we remove the share icon/button and instead allow the "•••" button to open the share sheet. We already have "Visit" as its own button anyways.

Not sure how we could implement this for now since "Share" technically means re-blog which in Calypso activates the re-blog popover.

> It also seems redundant to see the follow icon for sites I already follow.

Yup, that's been pointed out in the design review as well:

_Follow/Following buttons should not exist in the Following stream since you're following these sites already. They should only be in other streams/tags/lists._

kurzee commented:
>Yup, that's been pointed out in the design review as well:
>
>_Follow/Following buttons should not exist in the Following stream since you're following these sites already. They should only be in other streams/tags/lists._

Ah sorry! I missed that mention, @jancavan

aerych commented:
> I'd recommend we remove the share icon/button and instead allow the "•••" button to open the share sheet

The menu button provides other options besides sharing. It will also show unblocking, blocking, and unfollowing depending on the feed.  Its a little bare when it just shows a Share option which is one of the reasons it included Visit.

aerych commented:
> * button labels should be the same color as the icons themselves

The color was correct but there was a sneaky little text shadow lingering on the button’s label that I’ve removed.  Should look better now.


> * Align Visit button

Added a negative 2px inset to the button content.  It would be really great if the Gridicons font didn’t add padding around its glyphs (/rant) 


> Timestamp color should also be greyLighten10

Done


> In Calypso, Timestamps are in this format

We used to have the date in this format. However the mobile team decided to use a consistent date format within the app, specifically between the My Sites Post/Pages lists and the reader. The current format was chosen for better clarity for the user and with localization. Leaving this as is for now. If there is a compelling need for it to change we can revisit.


> Timestamp should link to the external site post

Keeping the current header behavior per later conversations.


> Follow/Following icon and labels can use a bit of space in between

This was a little weird. The title label had a left inset defined in interface builder but it was being ignored. I’ve opted to just handle it via code instead.


> Follow/Following buttons should not exist in the Following stream

Done.


> Doesn't seem like you can actually share/reblog the post?

Not quite sure what this means.  The share activity sheet appears and the post can be shared to any number of destinations.  Reblogging, however, is not currently a feature we support in the app.


> For the PostOptions ellipsis, we don't have Visit

Removed this option.


> default cards, images are 100px

Ah, good catch! Updated.


> Remove placeholder grav/blav

Updated to not show the placeholder unless one is defined.

aerych commented:
> The Follow icon in the top-right of the header seems crazy small and not obvious

I don't disagree, but I'm willing to see how our users receive the change.   Fwiw, the button is sized much larger than its icon so users should have no trouble tapping it.

aerych commented:
@kurzee @jancavan ready for another peek :)

kurzee commented on WordPress/Classes/ViewRelated/Reader/ReaderPostCardCell.swift:
```diff
@@ -253,7 +291,7 @@ import WordPressShared
         var url = featuredImageURL
         featuredImageDesiredWidth = featuredImageView.frame.width
-        let size = CGSize(width:featuredImageDesiredWidth, height:featuredMediaHeightConstraintConstant)
+        let size = CGSize(width:featuredImageDesiredWidth, height:100)
```
Maybe a `const` for `100`, or maybe it should read the `constant` value for the featured media height constraint?

kurzee commented:
New constraint issue:

```
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x608000281950 UIImageView:0x7fb0996406a0.width == 32   (active)>",
    "<NSLayoutConstraint:0x60000028a730 'UISV-hiding' UIImageView:0x7fb0996406a0.width == 0   (active)>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x608000281950 UIImageView:0x7fb0996406a0.width == 32   (active)>
```

That looks like it's the `avatarImageView` which is now being set to `hidden = true`. So, the width constraint itself would need a `999` priority instead of a `1000` to not break when being hidden. We'll have to run some additional testing for that as well, and ensure the `999` priority itself doesn't break the avatar if the header label text is super long.

kurzee commented:
>The menu button provides other options besides sharing. It will also show unblocking, blocking, and unfollowing depending on the feed. Its a little bare when it just shows a Share option which is one of the reasons it included Visit.

Hmm, maybe we could hide the ••• button when it will only end up presenting "Share"?

aerych commented:
> Maybe a const for 100

That was lazy of me. :/  I've changed this to just check the frame height.  

> New constraint issue:

Missed this in the console noise.  Rather than changing the priority we can just set aspect ratio and it should be solid.

aerych commented:
> Hmm, maybe we could hide the ••• button when it will only end up presenting "Share"?

No strong opinion about this either way. @jancavan what do you think?

kurzee commented on WordPress/Classes/ViewRelated/Reader/ReaderPostCardCell.swift:
```diff
@@ -291,7 +291,7 @@ import Gridicons
         var url = featuredImageURL
         featuredImageDesiredWidth = featuredImageView.frame.width
-        let size = CGSize(width:featuredImageDesiredWidth, height:100)
+        let size = CGSize(width:featuredImageDesiredWidth, height:featuredImageView.frame.height)
```
I'm not sure if the frame's height i reliable enough to use for fetching our desired layout's height. I think there are cases when this might be fired, prior to the final layout of the cell, resulting in a slightly incorrect frame height. Would be more reliable or guaranteed to derive the height from the constraint constant.

aerych replied:
It should be.  The frame is correct when the view awakes from its xib and since it no longer has a variable height, either for the iPad or when changing orientation we should be able to trust the frame.

jancavan commented:
@aerych Thanks for making all these changes! Everything else LGTM except for one minor thing:

Are the icons/labels the same color? They seem different to me. Should all be `greyLighten10` / `#a8bece`:

![screenshot 2016-11-18 11 21 00](https://cloud.githubusercontent.com/assets/4924246/20443341/5397f0a8-ad81-11e6-9da0-688018f53a4f.png)

> Added a negative 2px inset to the button content. It would be really great if the Gridicons font didn’t add padding around its glyphs (/rant)

Yeah, we've had to add custom positioning many times just to get them to align properly and it bloats the code.

kurzee commented:
Somewhat related: https://github.com/wordpress-mobile/WordPress-iOS/issues/6210

aerych commented:
> Are the icons/labels the same color

@jancavan, I'd have sworn they were ... but I missed where the labels were being set to greyDarken10.  Fixed now.  
The menu button looked a bit off as well.  It was previously being derived from a pdf asset.  I've changed it to be generated from a Gridicon like the share and visit icons.  How does it look now?  Any better?

Also, any thoughts about this? https://github.com/wordpress-mobile/WordPress-iOS/pull/6188#issuecomment-261542733

aerych commented:
@kurzee, I've restored the constraint wrangling for the featured image height.  Thanks for the extra testing and push back on this one. 
I'm leaving intact the featured image height for the iPad rather than reduce it to 100px.  This works better for now and we'll eventually use a different variant of the default card for the iPad that makes better use of the available size.  That forthcoming change should also resolve #6210.

kurzee commented:
@aerych everything is looking good! I'll keep testing if you make more changes, but you can :shipit: if ready.

jancavan commented:
@aerych 

Mentioned in https://github.com/wordpress-mobile/WordPress-iOS/pull/6188#issuecomment-261619539. That's supposed to also have the Follow/Unfollow button still.

aerych commented:
> Mentioned in #6188 (comment). That's supposed to also have the Follow/Unfollow button still

DoH!  Sorry. I completely missed that.  Fixing, tho users who are not signed in to wpcom will still just see a share button. Maybe we just look at that in a separate issue if there is a big concern there.

Also fixing an issue on the detail when a blavatar is nil. 

@kurzee game one last quick peek for paranoia sake and then we'll ship it?

kurzee commented:
@aerych everything looks good! :shipit:

aerych commented:
Thanks everyone! :)
